### PR TITLE
Upgrade tests for cinder volumes

### PR DESCRIPTION
**Pull request by abel-navarro**

### CODE REVIEW

**Comment by abel-navarro** [comment]
[(5) S56P10 upgrade: Attach volumes to the test-instances that are launched during the upgrade test jobs](https://trello.com/c/kRPfZFRM/338-5-s56p10-upgrade-attach-volumes-to-the-test-instances-that-are-launched-during-the-upgrade-test-jobs)

File: scripts/heat/ping_template.sh [hunk 1]
```diff
@@ -43,3 +43,18 @@ cat >> /usr/bin/ping_forever_neighbour << EOF
...
+  sudo mkfs -t ext4 -L cinder_volume /dev/vdb
+  sudo mount /dev/vdb /mnt
+
+  while [ 1 = 1 ]; do
+      date +%s >> /mnt/cinder_test.out
+      sleep 1
+  done
+EOF
+chmod +x /usr/bin/cinder_test
+/usr/bin/cinder_test
```
Comment (jsuchome) [inline scripts/heat/ping_template.sh]: Shoudln't this be executing in background?

Comment (abel-navarro) [inline scripts/heat/ping_template.sh]: right! fixing that

File: scripts/qa_crowbarsetup.sh [hunk 2]
```diff
@@ -4418,6 +4418,14 @@ function oncontroller_testpostupgrade
         scp cirros@$fip:/var/log/ping_outside.out ping_outside.$fip.out
         max=$(sed -n 's/^.* not available for: //p' ping_outside.$fip.out | sort -n | tail -n 1)
         echo "Maximum outage while pinging outside IP from $fip: $max seconds"
+
+        scp cirros@$fip:/mnt/cinder_test.out cinder_test.$fip.out
+        res=$(awk '$1!=p+1{print $1-p}{p=$1}' cinder_test.$fip.out | tail -n +2 | sort | uniq | tail -n 1)
```
Comment (jsuchome) [inline scripts/qa_crowbarsetup.sh]: That's a detail, but do you need `uniq` if you have the output already sorted and you are taking just the first item anyway?

File: scripts/heat/ping_template.sh [hunk 3]
```diff
@@ -43,3 +43,18 @@ cat >> /usr/bin/ping_forever_neighbour << EOF
 EOF
 chmod +x /usr/bin/ping_forever_neighbour
 /usr/bin/ping_forever_neighbour &
+
+cat >> /usr/bin/cinder_test << EOF
+  #!/bin/sh
+
+  sudo mkfs -t ext4 -L cinder_volume /dev/vdb
```
Comment (jsuchome) [inline scripts/heat/ping_template.sh]: This is just me not being used to work with volumes, but how are we sure that it's always `/dev/vdb`?

Comment (Itxaka) [inline scripts/heat/ping_template.sh]: I think that by default any *first* attached volume will get that by default for our kvm hypervisor.

I believe you can set the device name explicitly, but I think kvm ignores it anyway:

```
root@d52-54-77-77-77-01:~ # openstack server add volume --device /dev/vdg test1 test5
root@d52-54-77-77-77-01:~ # openstack volume list
+--------------------------------------+--------------+--------+------+--------------------------------+
| ID                                   | Display Name | Status | Size | Attached to                    |
+--------------------------------------+--------------+--------+------+--------------------------------+
| 52302393-8455-4436-86f6-c77701f6e439 | test5        | in-use |    1 | Attached to test1 on /dev/vdd  |

```

File: scripts/qa_crowbarsetup.sh (hunk 2 above)
Comment (Itxaka) [inline scripts/qa_crowbarsetup.sh]: Im having some difficulty trying to parse this into my head. We are trying to find if there is a gap of more than one second between lines in here right?

File: scripts/heat/ping_template.sh (hunk 3 above)
Comment (abel-navarro) [inline scripts/heat/ping_template.sh]: @jsuchome the cirros image has only one volume by default: `/dev/vda`. All my tests attaching a new (`cinder`) volume show it has been assigned to `/dev/vdb`. I think for our tests is a good enough solution.

@Itxaka you are right, even though you can name the device to use for the volume attachment, the cirros image will shamelessly ignore it.

File: scripts/qa_crowbarsetup.sh (hunk 2 above)
Comment (abel-navarro) [inline scripts/qa_crowbarsetup.sh]: Let me explain what this line does: the file is filled with a one timestamp per second (in seconds). The `awk` script checks if two consecutive timestamps have a difference of more than one second. If so, calculates the difference and writes that. If there are several gaps in the file, the script will write the difference for each of them.

After the `awk` script, we could consider the output is a list of gaps in seconds. For simplification, the first gap is always `1`, so we skip it with `tail -n +2`.

Then we sort the gaps by decreasing duration (`sort`) eliminating duplicates (`uniq`). We are just going to report the biggest gap, so we get the first in line (`tail -n 1`).

@jsuchome you are right, we might as well remove `uniq`. Will do that.

@Itxaka i hope my explanations covered your question.

**Comment by abel-navarro** [comment]
@Itxaka I reused the same script for the pings to simplify arguments passing to the VM. I would find it difficult to pass two files with heat templates (I don't know if it's possible). What it may make sense is to rename the file to clarify it is for scripts passing, not just for pings.

**Comment by abel-navarro** [comment]
I've been having some trouble getting 100% success for volumes attached to the instances via heat templates. Most of the times it works, but sometimes just one of the instances doesn't get the volume attachment.

I updated the script so it waits for the block device to be ready before starting the mount but it may happen that heat never attaches the volume at all. I would skip this case for now because I haven't found a reason for this to happen.

**Pull request merged**